src/Evolution/Systems/CurvedScalarWave/Worldtube/SingletonActions/ChangeSlabSize.hpp

```diff
@@ -0,0 +1,90 @@
+ * and adjust the worldtube slab size accordingly.
+ */
+struct ChangeSlabSize {
+  static constexpr size_t Dim = 3;
+  using inbox_tags = tmpl::list<
+      ::CurvedScalarWave::Worldtube::Tags::SphericalHarmonicsInbox<Dim>>;
+
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static Parallel::iterable_action_return_t apply(
+      db::DataBox<DbTagsList>& box, tuples::TaggedTuple<InboxTags...>& inboxes,
+      const Parallel::GlobalCache<Metavariables>& /*cache*/,
+      const ArrayIndex& /*array_index*/, ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) {
+    const auto& time_step_id = db::get<::Tags::TimeStepId>(box);
+    auto& inbox = tuples::get<Tags::SphericalHarmonicsInbox<Dim>>(inboxes);
+    if (inbox.empty()) {
+      return {Parallel::AlgorithmExecution::Retry, std::nullopt};
+    }
+    ASSERT(inbox.size() == 1,
+           "Received data from two different time step ids.");
+    const auto inbox_time_step_id = inbox.begin()->first;
+    const auto& inbox_slab = inbox_time_step_id.step_time().slab();
+    // we received data from a time step id with a different slab size,
+    // indicating a change in the global time step
+    if (inbox_time_step_id.step_time().slab() !=
+        time_step_id.step_time().slab()) {
+      ASSERT(inbox_time_step_id.slab_number() == time_step_id.slab_number(),
+             "Detected change in global time step but the new slab number is "
+             "different.");
+      ASSERT(inbox_slab.start() == time_step_id.step_time().slab().start(),
+             "The new slab should start at the same time as the old one.");
+      const auto new_step = inbox_slab.duration();
+      const auto new_next_time_step_id =
+          db::get<::Tags::TimeStepper<>>(box).next_time_id(inbox_time_step_id,
+                                                           new_step);
+      db::mutate<::Tags::Next<::Tags::TimeStepId>, ::Tags::TimeStep,
+                 ::Tags::Next<::Tags::TimeStep>, ::Tags::TimeStepId>(
+          make_not_null(&box),
+          [&new_next_time_step_id, &new_step, &inbox_time_step_id](
+              const gsl::not_null<TimeStepId*> next_time_step_id,
+              const gsl::not_null<TimeDelta*> time_step,
+              const gsl::not_null<TimeDelta*> next_time_step,
+              const gsl::not_null<TimeStepId*> local_time_step_id) {
+            *next_time_step_id = new_next_time_step_id;
+            *time_step = new_step;
+            *next_time_step = new_step;
+            *local_time_step_id = inbox_time_step_id;
+          });
+    }
+    const size_t expected_number_of_senders =
+        db::get<Tags::ElementFacesGridCoordinates<Dim>>(box).size();
+    if (inbox.at(time_step_id).size() < expected_number_of_senders) {
+      return {Parallel::AlgorithmExecution::Retry, std::nullopt};
```

Comment:
Generally messages should be waited for either by the action that uses them or, occasionally, by a dedicated action.  It seems odd for the `ChangeSlabSize` action to wait for messages that it is not using.  Can this be moved to the action that uses the data?